flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c

```diff
@@ -0,0 +1,500 @@
+		break;
+	case MSP_HEADER_M:
+		m->_state = b == '<' ? MSP_HEADER_SIZE : MSP_IDLE;
+		break;
+	case MSP_HEADER_SIZE:
+		m->_state = msp_state_size(m, b);
+		break;
+	case MSP_HEADER_CMD:
+		m->_state = msp_state_cmd(m, b);
+		break;
+	case MSP_FILLBUF:
+		m->_state = msp_state_fill_buf(m, b);
+		break;
+	case MSP_CHECKSUM:
+		m->_state = msp_state_checksum(m, b);
+		break;
+	case MSP_DISCARD:
+		m->_state = msp_state_discard(m, b);
+		break;
+	case MSP_MAYBE_UAVTALK2:
+		// e.g. 3c 20 1d 00
+		// second possible uavtalk byte
+		m->_state = (b&0xf0) == 0x20 ? MSP_MAYBE_UAVTALK3 : MSP_IDLE;
+		break;
+	case MSP_MAYBE_UAVTALK3:
+		// third possible uavtalk byte can be anything
+		m->_state = MSP_MAYBE_UAVTALK4;
+		break;
+	case MSP_MAYBE_UAVTALK4:
+		m->_state = MSP_IDLE;
+		// If this looks like the fourth possible uavtalk byte, we're done
+		if ((b & 0xf0) == 0) {
+			PIOS_COM_TELEM_RF = m->com;
+			return false;
+		}
+		break;
+	}
+
+	return true;
+}
+
+/**
+ * Module start routine automatically called after initialization routine
+ * @return 0 when was successful
+ */
+static int32_t uavoMSPBridgeStart(void)
+{
+	if (!module_enabled)
+		return -1;
+
+	if (FlightBatterySettingsHandle() != NULL
+			&& FlightBatteryStateHandle() != NULL) {
+		uint8_t currentPin;
+		FlightBatterySettingsCurrentPinGet(&currentPin);
+		if (currentPin != FLIGHTBATTERYSETTINGS_CURRENTPIN_NONE)
+			msp->msp_settings.use_current_sensor = true;
+		FlightBatterySettingsGet(&msp->msp_settings.battery_settings);
+		msp->msp_settings.batt_cell_count = msp->msp_settings.battery_settings.NbCells;
+	}
+	if (BaroAltitudeHandle() != NULL
+			&& PIOS_SENSORS_GetQueue(PIOS_SENSOR_BARO) != NULL)
+		msp->msp_settings.use_baro_sensor = true;
+
+	task = PIOS_Thread_Create(
+		uavoMSPBridgeTask, "uavoMSPBridge",
+		STACK_SIZE_BYTES, NULL, TASK_PRIORITY);
+	TaskMonitorAdd(TASKINFO_RUNNING_UAVOMSPBRIDGE,
+			task);
+
+	return 0;
+}
+
+/**
+ * Module initialization routine
+ * @return 0 when initialization was successful
+ */
+static int32_t uavoMSPBridgeInitialize(void)
+{
+	uint8_t module_state[MODULESETTINGS_ADMINSTATE_NUMELEM];
+	ModuleSettingsAdminStateGet(module_state);
+
+	if (pios_com_msp_id && (module_state[MODULESETTINGS_ADMINSTATE_UAVOMSPBRIDGE]
+			== MODULESETTINGS_ADMINSTATE_ENABLED)) {
+
+		msp = PIOS_malloc(sizeof(*msp));
+		if (msp != NULL) {
+			memset(msp, 0x00, sizeof(*msp));
+
+			msp->com = pios_com_msp_id;
+
+			PIOS_COM_ChangeBaud(msp->com, MSP_BAUDRATE);
+			module_enabled = true;
+			return 0;
+		}
+	}
+
+	module_enabled = false;
+
+	return -1;
+}
+MODULE_INITCALL(uavoMSPBridgeInitialize, uavoMSPBridgeStart)
+
+/**
+ * Main task routine
+ * @param[in] parameters parameter given by PIOS_Thread_Create()
+ */
+static void uavoMSPBridgeTask(void *parameters)
+{
+	while (1) {
+		uint8_t b = 0;
+		uint16_t count = PIOS_COM_ReceiveBuffer(msp->com, &b, 1, PIOS_QUEUE_TIMEOUT_MAX);
+		if (count) {
+			if (!msp_receive_byte(msp, b)) {
```

Comment:
Is it a good idea to just stop the task? Under what condition does that occur? It seems like the is an automatic UAVTalk detection, does that mean the same port will automatically switch to UAVTalk telemetry when it is connected to it? Does this work?